mlir/lib/Transforms/CompositePass.cpp

```diff
@@ -0,0 +1,106 @@
+//===- CompositePass.cpp - Composite pass code ----------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// CompositePass allows to run set of passes until fixed point is reached.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Transforms/Passes.h"
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_COMPOSITEFIXEDPOINTPASS
+#include "mlir/Transforms/Passes.h.inc"
+} // namespace mlir
+
+using namespace mlir;
+
+namespace {
+struct CompositeFixedPointPass final
+    : public impl::CompositeFixedPointPassBase<CompositeFixedPointPass> {
+  using CompositeFixedPointPassBase::CompositeFixedPointPassBase;
+
+  CompositeFixedPointPass(
+      std::string name_, llvm::function_ref<void(OpPassManager &)> populateFunc,
+      unsigned maxIterations) {
+    name = std::move(name_);
+    maxIter = maxIterations;
+    populateFunc(dynamicPM);
+
+    llvm::raw_string_ostream os(pipelineStr);
+    dynamicPM.printAsTextualPipeline(os);
+    os.flush();
+  }
+
+  LogicalResult initializeOptions(StringRef options) override {
+    if (failed(CompositeFixedPointPassBase::initializeOptions(options)))
+      return failure();
+
+    if (failed(parsePassPipeline(pipelineStr, dynamicPM))) {
+      llvm::errs() << "Failed to parse composite pass pipeline\n";
+      return failure();
+    }
+
+    return success();
+  }
+
+  LogicalResult initialize(MLIRContext * /*context*/) override {
+    if (maxIter <= 0) {
+      llvm::errs() << "Invalid maxIterations value: " << maxIter << "\n";
```

Comment:
(same)

Reply:
Switched to `emitError` here, but I don't think we have access to context in `initializeOptions` https://github.com/llvm/llvm-project/blob/6318dd82732c583fcfa12fe5b8d837e048dbb571/mlir/lib/Pass/PassRegistry.cpp#L43

Reply:
also, I can't move `parsePassPipeline` to `initialize` because `getDependentDialects` is called before `initialize`

Reply:
There isn’t a getContext() on the pass itself? (On a phone, I can’t look for a suggestion now)

in any case llvm::errs isn’t in scope of any pass or library code, so we need to figure out something else.

Reply:
I don't think context even exists as this point, but the calling function provides `errorHandler` callback: https://github.com/llvm/llvm-project/blob/6318dd82732c583fcfa12fe5b8d837e048dbb571/mlir/lib/Pass/PassRegistry.cpp#L41

We probably need to change `initializeOptions` signature to accept error callback as well.

Also, default pass implementation calls `parseFromString` which also diretly writes to `llvm::errs()`
https://github.com/llvm/llvm-project/blob/6318dd82732c583fcfa12fe5b8d837e048dbb571/mlir/lib/Pass/Pass.cpp#L53
https://github.com/llvm/llvm-project/blob/6318dd82732c583fcfa12fe5b8d837e048dbb571/mlir/lib/Pass/PassRegistry.cpp#L283

Reply:
> Also, default pass implementation calls parseFromString which also diretly writes to llvm::errs()

That's a problem as well, it was maybe not considered as critical because string parsing is more of a command-line integration point in general, but technically it still should work through injection (the kind of handler you're hinting at).

Reply:
Ok, I will look into updating `initializeOptions`.

Reply:
https://github.com/llvm/llvm-project/pull/87289

Reply:
Rebased and updated